src/dparse/parser.d

```diff
@@ -2100,7 +2120,7 @@ class Parser
      *     | $(RULE versionSpecification)
      *     ;)
      */
-    Declaration parseDeclaration(bool strict = false, bool mustBeDeclaration = false)
+    Declaration parseDeclaration(bool strict = false, bool mustBeDeclaration = false, bool inTemplateDeclaration = false)
```

Comment:
As far as I could tell from the implementation [1], AliasAssign is allowed only inside `template A(..){...}` declarations.

I understand your point, however, it is not possible to deduce whether the `static if` is at function level or at template level during parsing. In dmd, this is checked during semantic analysis, so I assume that the solution for libdparse is to create an AliasAssignDeclaration every time we encounter an assignment that occurs in non-function contexts. ParseDeclaration seems to be called at function scope level also. Any ideas on how I could differentiate between assignments at function level and assignments non-function levels?
 

[1] https://github.com/dlang/dmd/pull/11846/files?file-filters%5B%5D=.d&file-filters%5B%5D=.h#diff-90e6074482976dda59fff5b12d8aa29f1af3012b5db226a827e8e304b953ec10R6774

Reply:
I think you want to nest the inTemplateDeclaration parameter into all calls that should keep it, ie. here the parseConditionalDeclaration function and all the other `parse*Declaration` functions you want to support.

I personally think it would be more pretty doing this with some kind of parse state struct that is passed as argument in all related functions though if this is expected to expand in the future to more edge cases.

Reply:
I'm not sure this will work. Consider this case:

```d
template X(T)
{
    alias X = int;
    static if (cond)
        X = float;            // this one is currently accepted by dmd
    void fun()
    {
         X = T;               // while this one is not because the assignment
                                 // does not have the same parent as the declaration of X
    }
}
```

This is not going to get solved by passing the `inTemplate` boolean. Or maybe it could work, but that would mean inserting a primitive form of semantic analysis. Right now, libdparse issues an error when
assignments occur in non-function contexts. If we want to keep things in parser territory, we need to relax that condition and to create AliasAssignDeclarations if an assignment occurs in non-function context. Of course that will allow for code that is not an alias assignment to pass parsing, but that is fine.

Reply:
the parseFunctionDeclaration body would reset back the boolean value that you are in a template, (implicitly by simply not passing it down as parameter or for a struct explicitly) you will want to do that with anything adding scope.

Reply:
hmm, I'll try that out. Thanks for your assistance!